app/models/path_redirect.rb

```diff
@@ -0,0 +1,40 @@
+class PathRedirect < ApplicationRecord
+  SOURCES = %w[admin service].freeze
+
+  validates :old_path, presence: true, uniqueness: true
+  validates :new_path, presence: true
+
+  # Validate old_path != new_path
+  validates :old_path, exclusion: { in: ->(path_redirect) { [path_redirect.new_path] }, message: "the old_path cannot be the same as the new_path" }
+
+  validates :source, inclusion: { in: SOURCES }, allow_blank: true
+
+  # This issues a DB query so best to keep this validation as far down as possible
+  validate :new_path_wont_redirect
```

Comment:
If you want to eventually replace this with a DB unique index so you can avoid the DB query, you can do the following:

```sql
rhymes@/tmp:test> create extension if not exists hstore;
rhymes@/tmp:test> create unique index uniqut_ab on uniqt using btree (hstore(ARRAY[a,b], ARRAY[a,b]));
CREATE INDEX
Time: 0.009s
rhymes@/tmp:test> insert into uniqt values ('a', 'b');
INSERT 0 1
Time: 0.004s
rhymes@/tmp:test> insert into uniqt values ('b', 'a');
duplicate key value violates unique constraint "uniqut_ab"
DETAIL:  Key (hstore(ARRAY[a, b], ARRAY[a, b]))=("a"=>"a", "b"=>"b") already exists.
```

taken from https://stackoverflow.com/a/18361118/4186181

The reason why this works is because is because PostgreSQL is using its builtin key value store (hstore) to create an index where the permutation of `a, b` and `b, a` has to be unique.

In Rails terms it would translate into something like (untested):

```ruby
ActiveRecord::Base.connection.enable_extension("hstore")
ActiveRecord::Base.connection.execute(
  <<~SQL
    CREATE UNIQUE INDEX CONCURRENTLY "index_name"
    ON "page_redirects"
    USING btree (hstore(ARRAY[a,b], ARRAY[a,b]));
  SQL
)
```

in the migration file.

cc @vaidehijoshi

Reply:
TIL - this is _sweet_! Thanks for sharing 👍